### PR TITLE
add pointer(touch) event support

### PR DESCRIPTION
**Pull request by ssine**
along with window server SetSize opcode

### CODE REVIEW

File: src/emu/console/src/thread.cpp [hunk 1]
```diff
@@ -45,11 +45,41 @@ void set_mouse_down(void *userdata, const int button, const bool op) {
     emu->mouse_down[button] = op;
 }
 
+static eka2l1::drivers::input_event make_mouse_event_driver(float x, float y, int button, int action) {
```
Comment (pent0) [inline src/emu/console/src/thread.cpp]: Is it possible to make these parameter consts. They are not modified anywhere.

File: src/emu/console/src/thread.cpp [hunk 2]
```diff
@@ -45,11 +45,41 @@ void set_mouse_down(void *userdata, const int button, const bool op) {
...
+    eka2l1::drivers::input_event evt = *new eka2l1::drivers::input_event();
```
Comment (pent0) [inline src/emu/console/src/thread.cpp]: Don't do this. 

```suggestion
    eka2l1::drivers::input_event evt;
```

That leaks memory

File: src/emu/console/src/thread.cpp [hunk 3]
```diff
@@ -45,11 +45,41 @@ void set_mouse_down(void *userdata, const int button, const bool op) {
...
+ * \param button       0: left, 1: right, 2: other      
+ * \param action       0: press, 1: repeat(move), 2: release      
+ */
 static void on_ui_window_mouse_evt(void *userdata, eka2l1::point mouse_pos, int button, int action) {
+    float mouse_pos_x = static_cast<float>(mouse_pos.x), mouse_pos_y = static_cast<float>(mouse_pos.y);
+
     eka2l1::desktop::emulator *emu = reinterpret_cast<eka2l1::desktop::emulator *>(userdata);
-    float scale = emu->symsys->get_config()->ui_scale;
+    if (emu->symsys) {
+        float scale = emu->symsys->get_config()->ui_scale;
```
Comment (pent0) [inline src/emu/console/src/thread.cpp]: const

File: src/emu/console/src/thread.cpp [hunk 4]
```diff
@@ -45,11 +45,41 @@ void set_mouse_down(void *userdata, const int button, const bool op) {
...
+        mouse_pos_x /= scale;
+        mouse_pos_y /= scale;
+
+        auto mouse_evt = make_mouse_event_driver(mouse_pos_x, mouse_pos_y, button, action);
```
Comment (pent0) [inline src/emu/console/src/thread.cpp]: const

File: src/emu/debugger/src/imgui_debugger.cpp [hunk 5]
```diff
@@ -1047,6 +1047,12 @@ namespace eka2l1 {
 
             ImGui::SetNextWindowSize(ImVec2(size.x + 15.0f, size.y + 35.0f));
             ImGui::Begin(name.c_str());
+            const auto window_pos = ImGui::GetWindowPos();
+            const auto content_pos = ImGui::GetWindowContentRegionMin();
+            scr->absolute_pos_mtx.lock();
```
Comment (pent0) [inline src/emu/debugger/src/imgui_debugger.cpp]: I just thought this is nice.

File: src/emu/epoc/src/services/window/fifo.cpp [hunk 6]
```diff
@@ -72,6 +72,14 @@ namespace eka2l1::epoc {
                 break;
             }
 
+            case epoc::event_code::touch: {
+                // TODO: implement logics in 
+                // https://github.com/SymbianSource/oss.FCL.sf.os.graphics/blob/ff133bc50e6158bfb08cc093b0f0055321dcde99/windowing/windowserver/nga/SERVER/EVQUEUE.CPP#L630
+                // just purge it right now
+                q_.erase(q_.begin() + i);
```
Comment (pent0) [inline src/emu/epoc/src/services/window/fifo.cpp]: Nice too!

File: src/emu/epoc/src/services/window/window.cpp [hunk 7]
```diff
@@ -733,6 +734,57 @@ namespace eka2l1 {
         guest_evt_.key_evt_.repeats = 0;            // TODO?
     }
     
+    static std::array<bool, 3> mouse_button_pressed {};
```
Comment (pent0) [inline src/emu/epoc/src/services/window/window.cpp]: Static variable once modified, other modificiation will be invalid (on MSVC that's it, and on GCC its event more buggy). I highly suggest to rename this to button_pressed and move this into window_server along with the other function.

File: src/emu/console/src/thread.cpp (hunk 4 above)
Comment (ssine) [reply src/emu/console/src/thread.cpp]: mouse_evt was passed to queue_input_from_driver next line which doesn't accept const parameter

**Pull request merged**